stdlib/string.gr

```diff
@@ -519,3 +519,596 @@ export let endsWith = (p: String, s: String) => {
     Memory.compare(p, s + n - m, m) == 0n
   }
 }
+
+export enum Encoding {
+  UTF8,
+  UTF16_BE,
+  UTF16_LE,
+  UTF32_BE,
+  UTF32_LE,
+}
+
+// String->Byte encoding and helper functions:
+
+@disableGC
+let utf16Length = (s: String) => {
+  let (>>>) = WasmI32.shrU
+  let (<<) = WasmI32.shl
+  let (+) = WasmI32.add
+  let (&) = WasmI32.and
+  let (<) = WasmI32.ltU
+  let (==) = WasmI32.eq
+
+  let size = WasmI32.fromGrain(byteLength(s)) >>> 1n
+  let len = WasmI32.fromGrain(length(s)) >>> 1n
+
+  let s = WasmI32.fromGrain(s)
+
+  let mut ptr = s + 8n
+  let end = ptr + size
+  let mut size = 0n // <- number of UTF-16 code words
+
+  while (ptr < end) {
+    let byte = WasmI32.load8U(ptr, 0n)
+    let n = if ((byte & 0x80n) == 0x00n) {
+      1n
+    } else if ((byte & 0xF0n) == 0xF0n) {
+      4n
+    } else if ((byte & 0xE0n) == 0xE0n) {
+      3n
+    } else {
+      2n
+    }
+    if (n == 4n) {
+      size += 2n
+    } else {
+      size += 1n
+    }
+    ptr += n
+  }
+  // multiply by two to get number of bytes (equivalent to left-shift)
+  tagSimpleNumber(size << 1n)
+}
+
+let encodedLength = (s: String, encoding) => {
+  match(encoding) {
+    UTF32_BE => length(s) * 4,
+    UTF32_LE => length(s) * 4,
+    UTF16_BE => utf16Length(s),
+    UTF16_LE => utf16Length(s),
+    UTF8 => byteLength(s),
+  }
+}
+
+exception MalformedUnicode
+
+@disableGC
+let getCodePoint = (ptr: WasmI32) => {
```

Comment:
This already exists in the Char module. I wonder if it's worth trying to share them.

Reply:
Per discussion offline, deferring to #684.